Добавлена возможность работать с трекером задач через HTTP

PrettyDG opened the pull request:
К сожалению, не удалось реализовать один из тестов по добавлению Epic'а - testAddEpic. Во время его преобразования в json возникает ошибка с тем, что "localDateTime = null", т.к. эпик состоит только из названия и описания. Не получилось найти адекватное решение данной проблемы

PrettyDG committed:
Добавлена возможность работать с трекером задач через HTTP

PrettyDG committed:
Визуальные изменения - 1

avfyodorov commented on src/http/handlers/BaseHttpHandler.java:
```diff
@@ -0,0 +1,61 @@
+        }
+    }
+
+    static class LocalDateTimeTypeAdapter extends TypeAdapter<LocalDateTime> {
```
Лучше было бы оформить эти адаптеры как отдельные классы. Тогда не надо было бы их пересоздавать в тестах. 

На усмотрение.

avfyodorov commented on src/http/handlers/BaseHttpHandler.java:
```diff
@@ -0,0 +1,61 @@
+
+public class BaseHttpHandler {
+
+    protected static Gson gson = new GsonBuilder()
```
Так же хорошо было бы вынести  создание Gson  в  метод в класс-Managers, так как он является своего рода фабрикой, отвечающей за создание подобных объектов). Лучше бы объект создать один раз в Managers, а потом переиспользовать.
Этот объект достаточно тяжелый.

avfyodorov commented on src/http/handlers/EpicsHandler.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
```
public void handle(HttpExchange exchange) throws IOException {

Логично было бы поместить этот метод в базовом классе, написать заглушки-обработчики, а в классах наследников переопределять только эти методы (handleGet, handlePost….). Такое решение лучше соответствует принципам OOP 
На усмотрение.

avfyodorov commented on test/HttpTaskServerTest.java:
```diff
@@ -0,0 +1,213 @@
+
+//    @Test
+//    public void testAddEpic() throws IOException, InterruptedException {
+//        Epic epic = new Epic("Epic1", "Epic1 description");
```
А если задать Эпику какие-то начальные (по-умолчанию) значения продолжительности и времён.

avfyodorov reviewed (commented):
Добрый вечер, Данила! 
Хорошая работа, от меня пара предложений, они на Ваше усмотрение.

PrettyDG committed:
Adapter'ы перемещены в отдельную утилитарную папку.
Перенесено создание Gson в метод Managers, предварительно проверяется, не был ли он уже создан.
Метод 'handle' был перенесён в BaseHttpHandler.
Была решена проблема с созданием Эпика в тестах.

avfyodorov commented on src/http/handlers/BaseHttpHandler.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+
+    public abstract Object handleGet(HttpExchange exchange) throws IOException;
```
Здесь лучше было бы реализовать не абстрактные методы, а вот такие заглушки. Тогда в классе-наследнике не придётся переписывать все методы, а только те, которые нужны по бизнес логике. 


 public void handlePost(HttpExchange exchange) throws IOException {
        writeToUser(exchange, "Данный метод не предусмотрен");
    }

 
    public void handleDelete(HttpExchange exchange) throws IOException {
        writeToUser(exchange, "Данный метод не предусмотрен");
    }

avfyodorov reviewed (commented):
Добрый день, Данила!

Замечаний нет.
Работа принята.

Оставил комментарий по поводу реализации обработчика, посмотрите, пожалуйста. 

Есть просьба-оставлять хоть какую-то реакцию на мои комментарии. Сделано/ Не сделано, потому что не согласен/ Не заметил. 🙂
